docs/introduction-to-smart-contracts.rst

```diff
@@ -432,6 +434,12 @@ you should minimize what you store in persistent storage to what the contract ne
 Store data like derived calculations, caching, and aggregates outside of the contract.
 A contract can neither read nor write to any storage apart from its own.
 
+Similar to storage, there is another data area called **transient storage**, which main difference
+is the fact that it is reset at the end of each transaction. The values stored in this data location
+are persisted across function calls originating from the first call of the transaction. When the transaction
+ends, the transient storage is reset and all values are lost to subsequent transaction calls. Albeit that,
+the cost of reading and writing to transient storage is significantly cheaper than storage.
```

Comment:
```suggestion
Similar to storage, there is another data area called **transient storage**, where the main difference
is the fact that it is reset at the end of each transaction.
The values stored in this data location persist only across function calls originating from the first call of the transaction.
When the transaction ends, the transient storage is reset and the values stored there become unavailable to calls in subsequent transactions.
Albeit that, the cost of reading and writing to transient storage is significantly lower than for storage.
```

Reply:
I think that we should include some warning about the consequences of not resetting transient storage. Essentially what we warned about in the blog post. Not sure if this is the best place - it's just an intro after all - but it should be somewhere. Perhaps we should keep it short here but have something more detailed elsewhere where we properly discuss data locations?

We could also use an example - perhaps the reentrancy lock should be one of the examples for beginners?

Reply:
I will add a brief description of the problem and reference another (to be created) section where transient storage is explored further.
I think the example is a good idea, but should we add it here or in another section?

Reply:
I added a section briefly describing transient storage and showing a simple reentrancy lock example.

Reply:
> I think the example is a good idea, but should we add it here or in another section?

I was thinking about putting one in the "Solidity by Example" section. But maybe having it in the section about transient storage is enough.